contributors/design-proposals/resource-management/device-plugin.md

```diff
@@ -340,37 +325,39 @@ message Device {
 We want Kubelet as well as the Device Plugins to recover from failures
 that may happen on any side of this protocol.
 
-At the communication level, gRPC is a very strong piece of software and
-is able to ensure that if failure happens it will try it's best to recover
-through exponential backoff reconnection and Keep Alive checks.
-
 The proposed mechanism intends to replace any device specific handling in
-Kubelet. Therefore in general, device plugin failure or upgrade means that
-Kubelet is not able to accept any pod requesting a Device until the upgrade
-or failure finishes.
+Kubelet. Therefore in general, device plugin failure means that
+Kubelet is not able to accept any new pod requesting a Device until the
+failure finishes.
 
 If a device fails, the Device Plugin should signal that through the
-`ListAndWatch` gRPC stream. We then expect Kubelet to fail the Pod.
+`ListAndWatch` gRPC stream. We then expect Kubelet to avoid scheduling pods on the
+failed device.
 
 If any Device Plugin fails the behavior we expect depends on the task Kubelet
 is performing:
-* In general we expect Kubelet to remove any devices that are owned by the failed
-  device plugin from the node capacity. We also expect node allocatable to be
-  equal to node capacity.
-* We however do not expect Kubelet to fail or restart any pods or containers
-  running that are using these devices.
+* Kubelet will remove any devices that are owned by the failed
+  device plugin from the node capacity and allocatable. This will prevent future pods from being scheduled on the node.
+* We however do not expect Kubelet to fail or restart any running pods or containers
+  that are using these devices. This is supported through two mechanisms:
+  First, Kubelet caches and checkpoints device runtime information to avoid calling device plugin Allocate for devices already allocated to a Pod.
+  Second, we expose a PluginResourceUpdater from ContainerManager and pass it to lifecycle.NewPredicateAdmitHandler().
+  predicateAdmitHandler calls this function after it gets nodeInfo and before calling GeneralPredicates.
+  In PluginResourceUpdater, we use the cached device allocation information to update NodeInfo.allocatableResource
+  to at least equal to in-use capacity. This avoids predicateAdmit failure due to allocatable device plugin
+  resource dropping to zero.
```

Comment:
There's an implicit case here unless I missed it in the text: when a container terminates and was using a device advertised by a failed plugin, the same mechanism decreases the allocatable quantity for that resource so that a subsequent pod cannot use it. Could we document that scenario here too?

Reply:
I added a couple sentences to clarify this. PTAL.